core/src/air/word_64.rs

```diff
@@ -0,0 +1,145 @@
+use core::fmt::Debug;
+use std::{
+    array::IntoIter,
+    ops::{Index, IndexMut},
+};
+
+use p3_air::AirBuilder;
+use p3_field::{AbstractField, Field};
+use serde::{Deserialize, Serialize};
+use sphinx_derive::AlignedBorrow;
+
+use super::BaseAirBuilder;
+use crate::air::Word;
+
+/// The size of a word64 in bytes.
+pub const WORD64_SIZE: usize = 8;
+
+/// A double word is a 64-bit value represented in an AIR.
+#[derive(
+    AlignedBorrow, Clone, Copy, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize,
+)]
+#[repr(C)]
+pub struct Word64<T>(pub [T; WORD64_SIZE]);
+
+impl<T> Word64<T> {
+    /// Applies `f` to each element of the word64.
+    pub fn map<F, S>(self, f: F) -> Word64<S>
+    where
+        F: FnMut(T) -> S,
+    {
+        Word64(self.0.map(f))
+    }
+
+    /// Extends a variable to a word64.
+    pub fn extend_var<AB: BaseAirBuilder<Var = T>>(var: T) -> Word64<AB::Expr> {
+        Word64([
+            AB::Expr::zero() + var,
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+        ])
+    }
+}
+
+impl<T: Clone> Word64<T> {
+    /// Splits into two words.
+    pub fn to_le_words(self) -> [Word<T>; 2] {
+        let limbs: Vec<T> = self.into_iter().collect();
+        [
+            limbs[..4].iter().cloned().collect(),
+            limbs[4..].iter().cloned().collect(),
+        ]
+    }
+}
+
+impl<T: AbstractField> Word64<T> {
+    /// Extends a variable to a word64.
+    pub fn extend_expr<AB: BaseAirBuilder<Expr = T>>(expr: T) -> Word64<AB::Expr> {
+        Word64([
+            AB::Expr::zero() + expr,
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+        ])
+    }
+
+    /// Returns a word64 with all zero expressions.
+    pub fn zero<AB: BaseAirBuilder<Expr = T>>() -> Word64<T> {
+        Word64([
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+            AB::Expr::zero(),
+        ])
+    }
+}
+
+impl<F: Field> Word64<F> {
+    /// Converts a word64 to a u64.
+    pub fn to_u64(&self) -> u64 {
+        // TODO: avoid string conversion
+        u64::from_le_bytes(self.0.map(|x| x.to_string().parse::<u8>().unwrap()))
+    }
+}
+
+impl<V: Copy> Word64<V> {
+    /// Reduces a word64 to a single variable.
+    pub fn reduce<AB: AirBuilder<Var = V>>(&self) -> AB::Expr {
+        let base = [1, 1 << 8, 1 << 16, 1 << 24].map(AB::Expr::from_canonical_u32);
```

Comment:
Not sure if this is the right behavior since it's ignoring the high limbs

Reply:
It's not, and for a `Word64` the `reduce` probably makes not that much sense compared to a `Word`, since it's almost a guarantee it will overflow for the higher limbs. I might just remove this function since I think it is not being used